src/dbnode/storage/index/block.go

```diff
@@ -823,30 +852,235 @@ func (b *block) addQueryResults(
 	results BaseResults,
 	batch []doc.Document,
 ) ([]doc.Document, int, error) {
-	// Checkout the lifetime of the query before adding results
+	// checkout the lifetime of the query before adding results.
 	queryValid := cancellable.TryCheckout()
 	if !queryValid {
-		// Query not valid any longer, do not add results and return early
+		// query not valid any longer, do not add results and return early.
 		return batch, 0, errCancelledQuery
 	}
 
-	// Try to add the docs to the resource
+	// try to add the docs to the resource.
 	size, err := results.AddDocuments(batch)
 
-	// Immediately release the checkout on the lifetime of query
+	// immediately release the checkout on the lifetime of query.
 	cancellable.ReleaseCheckout()
 
-	// Reset batch
+	// reset batch.
 	var emptyDoc doc.Document
 	for i := range batch {
 		batch[i] = emptyDoc
 	}
 	batch = batch[:0]
 
-	// Return results
+	// return results.
 	return batch, size, err
 }
 
+// Aggregate acquires a read lock on the block so that the segments
+// are guaranteed to not be freed/released while accumulating results.
+// This allows references to the mmap'd segment data to be accumulated
+// and then copied into the results before this method returns (it is not
+// safe to return docs directly from the segments from this method, the
+// results datastructure is used to copy it every time documents are added
+// to the results datastructure). This is similar to how Query() operates.
+// NB: Aggregate is an optimization of the general aggregate Query approach
+// for the case when we can skip going to raw documents, and instead rely on
+// pre-aggregated results via the FST underlying the index.
+func (b *block) Aggregate(
+	cancellable *resource.CancellableLifetime,
+	opts QueryOptions,
+	results AggregateResults,
+) (bool, error) {
+	b.RLock()
+	defer b.RUnlock()
+
+	if b.state == blockStateClosed {
+		return false, ErrUnableToQueryBlockClosed
+	}
+
+	aggOpts := results.AggregateResultsOptions()
+	iterateTerms := aggOpts.Type == AggregateTagNamesAndValues
+	iterateOpts := fieldsAndTermsIteratorOpts{
+		iterateTerms: iterateTerms,
+		allowFn: func(field []byte) bool {
+			// skip any field names that we shouldn't allow.
+			if bytes.Equal(field, doc.IDReservedFieldName) {
+				return false
+			}
+			return aggOpts.TermFilter.Allow(field)
+		},
+	}
+
+	iter, err := b.newFieldsAndTermsIteratorFn(nil, iterateOpts)
+	if err != nil {
+		return false, err
+	}
+
+	var (
+		size       = results.Size()
+		batch      = b.opts.AggregateResultsEntryArrayPool().Get()
+		batchSize  = cap(batch)
+		iterClosed = false // tracking whether we need to free the iterator at the end.
```

Comment:
nah so it's like the standard cleanup pattern we use in a few places

```
object := ...
cleanedUp := false

defer func() {
  if !cleanedUp {
    object.Cleanup()
  }
}()

// do thing 1 for object
// if it fails, just early `return`

// so on

// finally, 
cleanedUp = true
if err :=object.Cleanup(); err != nil {
  return err
}
```

this way we can be sure all exit paths from the function cleanup the object; either because we manually do it; or the defer takes care of it. 

without the defer (and the var). i'd have to interlace the cleanup calls at every exit point from the function. which seems brittle at best.